pkg/istio-agent/xds_proxy.go

```diff
@@ -119,14 +123,48 @@ func initXdsProxy(ia *Agent) (*XdsProxy, error) {
 		LocalHostAddr: localHostAddr,
 	}
 	proxy := &XdsProxy{
-		istiodAddress:  ia.proxyConfig.DiscoveryAddress,
-		clusterID:      ia.secOpts.ClusterID,
-		localDNSServer: ia.localDNSServer,
-		stopChan:       make(chan struct{}),
-		healthChecker:  health.NewWorkloadHealthChecker(ia.proxyConfig.ReadinessProbe, envoyProbe),
-		xdsHeaders:     ia.cfg.XDSHeaders,
-		xdsUdsPath:     ia.cfg.XdsUdsPath,
-		wasmCache:      wasm.NewLocalFileCache(constants.IstioDataDir, wasm.DefaultWasmModulePurgeInteval, wasm.DefaultWasmModuleExpiry),
+		istiodAddress: ia.proxyConfig.DiscoveryAddress,
+		clusterID:     ia.secOpts.ClusterID,
+		handlers:      map[string]ResponseHandler{},
+		stopChan:      make(chan struct{}),
+		healthChecker: health.NewWorkloadHealthChecker(ia.proxyConfig.ReadinessProbe, envoyProbe),
+		xdsHeaders:    ia.cfg.XDSHeaders,
+		xdsUdsPath:    ia.cfg.XdsUdsPath,
+		wasmCache:     wasm.NewLocalFileCache(constants.IstioDataDir, wasm.DefaultWasmModulePurgeInteval, wasm.DefaultWasmModuleExpiry),
+	}
+
+	if ia.localDNSServer != nil {
+		proxy.handlers[v3.NameTableType] = func(resp *discovery.DiscoveryResponse) error {
+			if len(resp.Resources) == 0 {
+				return nil
+			}
+			var nt nds.NameTable
+			if err := ptypes.UnmarshalAny(resp.Resources[0], &nt); err != nil {
+				log.Errorf("failed to unmarshall name table: %v", err)
+				return err
+			}
+			ia.localDNSServer.UpdateLookupTable(&nt)
+			return nil
+		}
+	}
+	if ia.secretCache != nil {
+		proxy.handlers[v3.ProxyConfigType] = func(resp *discovery.DiscoveryResponse) error {
+			if len(resp.Resources) == 0 {
+				return fmt.Errorf("empty response")
+			}
+			var pc meshconfig.ProxyConfig
+			if err := gogotypes.UnmarshalAny(gogo.ConvertAny(resp.Resources[0]), &pc); err != nil {
+				log.Errorf("failed to unmarshall proxy config: %v", err)
+				return err
+			}
+			caCerts := pc.GetCaCertificatesPem()
+			log.Debugf("received new certificates to add to mesh trust domain: %v", caCerts)
+			trustBundle := []byte{}
+			for _, cert := range caCerts {
+				trustBundle = util.AppendCertByte(trustBundle, []byte(cert))
+			}
+			return ia.secretCache.UpdateConfigTrustBundle(trustBundle)
```

Comment:
do we want to do a diff in UpdateConfigTrustBundle and not push to envoy if nothing changed? Two small concerns if not:
* Istiod disconnects every 30 minutes. We already get tons of questions about these logs, now they will be worse:
```
2021-02-25T16:05:44.402385Z     info    xdsproxy        connected to upstream XDS server: localhost:15012
2021-02-25T16:05:44.457012Z     info    ads     XDS: Incremental Pushing:0 ConnectedEndpoints:2 Version:
2021-02-25T16:05:44.457108Z     info    cache   returned workload trust anchor from cache       ttl=23h58m58.542898088s
2021-02-25T16:05:44.457136Z     info    sds     SDS: PUSH       resource=ROOTCA
```
* Envoy SDS has tons of bugs, pushing new rootca every 30 minutes might trigger some. maybe I am paranoid though...

Reply:
Will fix this in the following PR